hasOne association could not restore

JeskTop opened the pull request:
When I find @pobocks fix the problem https://github.com/radar/paranoia/pull/119 , and I use the new paranoia and find another bug. 
The problem can look at my first commit what is a test and the problem like https://github.com/radar/paranoia/issues/120 .

JeskTop committed:
add two test unit and could not pass

JeskTop committed:
pass test and fix the has one could not restore problem

radar commented on lib/paranoia.rb:
```diff
@@ -114,6 +114,10 @@ def restore_associated_records
       end
+
+      if association_data.nil? && association.macro.to_s == "has_one"
+        Object.const_get(association.name.to_s.camelize).only_deleted.where("#{self.class.name.to_s.underscore}_id", self.id).first.try(:restore, recursive: true)
```
Um, wow.

Could you please write some documentation about this? There's a lot of meta programming that goes on within this one line!

JeskTop replied:
Hi @radar , my English is not very good, write to the following several documentations, do not know whether or not clear?

JeskTop commented:
My commits fix the problem is the `has_one` counld not restore. 
I submitted before the test is passed, because the destroy object did not find from the database and it just a catch or not. Code: https://github.com/radar/paranoia/blob/rails4/test/paranoia_test.rb#L382

JeskTop commented:
How can I find the has_one object?

The difference between `has_one` and `has_many` like this:

``` ruby
# User has_many Message
@user.messages.where(title: "abc")   #This is can work!
```

``` ruby
# User has_one Role
@user.role.where(name: "abc")  #This is can not work!
```

So if I want to find the `role` must use:

``` ruby
Role.where(name: "abc", user_id: @user.id)
```

It just like:

``` ruby
Object.const_get("role").where(name: "abc", user_id: @user.id)
```

So this one line:

``` ruby
Object.const_get(association.name.to_s.camelize).only_deleted.where("#{self.class.name.to_s.underscore}_id", self.id).first.try(:restore, recursive: true)
```

This mean is use `association.name.to_s.camelize` find the `has_one` model name and use `self.class.name.to_s.underscore` to setting this class `id` (`user_id`).

And then user `User has_one Role` it will work look like:

``` ruby
Object.const_get("role").only_deleted.where("user_id", self.id).first.try(:restore, recursive: true)
```

Or

``` ruby
Role.only_deleted.where("user_id", self.id).first.try(:restore, recursive: true)
```

pobocks commented:
Is this guaranteed to be the fkey name in question?

``` ruby
{self.class.name.to_s.underscore}_id
```

It seems like there'd need to be a check for key columns specified in the options.

JeskTop commented:
@pobocks Oh. Yes! I must think about how to use the `foreign_key` to find the model.

JeskTop committed:
Merge branch 'rails4' into setting_restore

JeskTop committed:
use classs_name and foreign_key restore

when the model has_one use class_name or foreign_key, restore use
class_name or setting the foreign_key.

JeskTop commented:
Hey, @pobocks. I pushed the new commit and checked the specified `foreign_key` and `class_name`.

adityashedge commented:
I'm having an issue with restoring dependent has_one relation recursively in rails 4.1. Can you check if this merge request fixes it and if it does please merge it ASAP. Thank  you in advance.

Exact Problem:
User has_one UserInfo and UserInfo belongs_to User.
Unable to restore UserInfo when restoring user and all dependent records recursively.
UserInfo is soft deleted but not restored when restoring User recursively.

JeskTop commented:
@adityashedge 
Thanks. I think the commit can fix all of your problems.
You can see my test. If `User has_one UserInfo and UserInfo belongs_to User`, It will restore `UserInfo` when restoring user and all dependent records recursively.
And if you find the another problem, you can tell me.

willsu commented:
+1 @JeskTop This fixed my issue with `restore(recursive: true)` on a has_one association.

@radar Any chance of getting this pull request accepted? I'm pointing my Gemfile at the @JeskTop fork and `setting_restore` branch for the time being.

radar commented:
The PR no longer applies cleanly to the targeted branch. If this PR got an update to fix that, then I would gladly pull this in.

JeskTop committed:
merge rails4 branch and fix conflict

JeskTop commented:
What do you mean clean? I alreay update the `rails4` branch and merge for my `setting_restore` branch. If @radar has another problems, you can tell me and I will fix that. Thanks.

ccaratori commented:
Hey guys is this going to be merged and released any time soon?

radar commented:
@JeskTop I still can't automatically merge this PR, and it has failing tests.

![screen shot 2014-09-15 at 11 18 38 am](https://cloud.githubusercontent.com/assets/2687/4266181/3cc0ca4a-3c76-11e4-9d51-6145e9128801.png)

Once those failing tests and merge conflicts are fixed, then I can look at this again.

JeskTop committed:
Merge branch 'rails4' into setting_restore

Conflicts:
	test/paranoia_test.rb

JeskTop commented:
@radar Thanks. I already fix all conflicts. And push the commit.

radar commented:
Thank you very much @JeskTop.
